inference-engine/src/transformations/src/transformations/convert_precision.cpp

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/convert_precision.hpp"
+
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset3.hpp>
+
+using namespace ngraph;
+
+bool ngraph::pass::ConvertPrecision::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    static std::map<ngraph::NodeTypeInfo, std::function<bool(std::shared_ptr<Node>, element::Type, size_t idx)>> type_to_fuse {
+        {opset3::ShapeOf::type_info, fuse_type_to_shapeof},
+        {opset3::Convert::type_info, fuse_type_to_convert},
+        {opset3::Constant::type_info, fuse_type_to_constant},
+        {opset3::Parameter::type_info, fuse_type_to_parameter},
+    };
+    // TODO: add nodes from body
+    for (auto &node : f->get_ordered_ops()) {
+        node->validate_and_infer_types();
+
+        for (auto output : node->outputs()) {
+            if (output.get_element_type() == m_from) {
+                // If node type in map and convert can be fused into node we skip Convert creation
+                if (type_to_fuse.count(node->get_type_info()) &&
+                    type_to_fuse.at(node->get_type_info())(node, m_to, output.get_index())) {
+                    continue;
+                }
+                // Create Convert operation and reconnect consumers
+                auto consumers = output.get_target_inputs();
+                auto convert = std::make_shared<opset3::Convert>(output, m_to);
+                for (auto & input : consumers) {
+                    input.replace_source_output(convert);
+                }
+            }
+        }
+    }
+
+    // Check that function has no more operations with output type equal to m_to
+    for (auto &node : f->get_ordered_ops()) {
+        for (auto &output : node->outputs()) {
+        }
+    }
+}
+
+bool fuse_type_to_shapeof(std::shared_ptr<Node> node, element::Type to, size_t idx) {
+    if (auto shapeof = as_type_ptr<opset3::ShapeOf>(node)) {
+        if (to == element::i32 || to == element::i64) {
+            shapeof->set_output_type(to);
+            return true;
+        }
+    }
+    return false;
+}
+
+bool fuse_type_to_parameter(std::shared_ptr<Node> node, element::Type to, size_t idx) {
+    if (auto param = as_type_ptr<opset3::Parameter>(node)) {
+        param->set_element_type(to);
+    }
+    return false;
+}
+
+bool fuse_type_to_convert(std::shared_ptr<Node> node, element::Type to, size_t idx) {
+    if (auto convert = as_type_ptr<opset3::Convert>(node)) {
+        convert->set_convert_element_type(to);
+    }
+    return false;
+}
+
+template <element::Type_t PREC_FROM, element::Type_t PREC_TO>
+std::shared_ptr<Node> change_constant_precision(std::shared_ptr<opset3::Constant> constant) {
+    using src_type = typename element_type_traits<PREC_FROM>::value_type;
+    using dst_type = typename element_type_traits<PREC_TO>::value_type;
+    return std::make_shared<ngraph::opset3::Constant>(PREC_TO, constant->get_shape(), constant->cast_vector<dst_type>());
```

Comment:
CC @AndrewBakalinIntel 

We need to make sure that `cast_vector` does not have the same problem which us solved by saturation case

Reply:
Actually it has.
@GlebKazantaev please take a look at this https://github.com/openvinotoolkit/openvino/pull/1304
The main problem that I tried to solve by introducing `saturate_cast` is the conversion of `int64_t::max()` to `int32_t::max()` in StridedSlice for Faster-RCNN (`int64_t::max()` is used as a value for `end` attribute in original onnx model). In `static_cast` this conversion is unspecified so the plugins which are not support I64 will receive incorrect attribute values.

Reply:
Oh yes I know :) I just submitted this draft to check the CI models. But thanks anyway!

Reply:
Done.